### PR TITLE
Add `external-to-ref` pass

### PR DESCRIPTION
**Pull request by nathanielnrn**
This PR introduces `external-to-ref` which takes all cells with the attributes `@external(1)` and removes the attributes while denoting them `ref` cells. This pass goes through every module in a given calyx program (as opposed to just the toplevel), which I think is what we want?

This PR also changes the `well-formed` pass to allow for `ref` cells in toplevels, which previously wasn't allowed.

Some things to note/questions I am not sure about:

1. This pass changes every cell, not just memory cells/toplevel cells. I think this is ok but would like to confirm.
2. The only registering I did for this pass was add it to `default-passes.rs`. Is that all that needs to be done? I followed the [instructions about adding a new pass](https://docs.calyxir.org/new-pass.html), but want to check that those instructions are not missing anything.

### CODE REVIEW

File: calyx-opt/src/passes/external_to_ref.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        Self: Sized,
+    {
+        let external_to_ref = ExternalToRef;
+        Ok(external_to_ref)
+    }
+
+    fn clear_data(&mut self) {
+        //TODO: do we need this?
+    }
+}
```
Comment (nathanielnrn) [inline calyx-opt/src/passes/external_to_ref.rs:32]: Confession: This was entirely pattern matched from other passes (I think `externalize`) and I do not understand what it is doing. Happy to hear about any changes that should be made.

File: calyx-opt/src/passes/external_to_ref.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+use crate::traversal::{Action, ConstructVisitor, Named, VisResult, Visitor};
+use calyx_ir::{self as ir, GetAttributes, LibrarySignatures};
+use calyx_utils::CalyxResult;
+
+//Turns memory cell primitives with the `@external(1)` attribute into
+// `ref` memory cells without the `@external` attribute.
```
Comment (sampsyo) [inline calyx-opt/src/passes/external_to_ref.rs:6]: Perhaps you want this to be a Rust documentation comment. Those are indicated with triple-slashes, `///`.

File: calyx-opt/src/passes/external_to_ref.rs (hunk 1 above)
Comment (sampsyo) [reply calyx-opt/src/passes/external_to_ref.rs:32]: I'm afraid I don't have the answer to this one—perhaps somebody else knows whether this is required. It sure seem like it is required for every pass to me; I would feel comfortable assuming that this is exactly what you need.

File: tests/passes/well-formed/main-ref-cell.expect
Comment (sampsyo) [inline tests/passes/well-formed/main-ref-cell.expect:1]: Let's just remove this test. IMO there is nothing important left to test here: this was meant to check that an error was triggered when there was a `ref` in `main`, but now we don't want anything special to happen in that case.

**Review by sampsyo — COMMENTED** [review]
Excellent! This all looks great! Thanks for even including a test. :tada: This seems like the right thing to me!

I'm also tagging @ayakayorihiro here, because her FIRRTL work has relied on a `sed`-based alternative to this so far. Perhaps she can use this pass instead!

Some answers to the questions above:

> This pass changes every cell, not just memory cells/toplevel cells. I think this is ok but would like to confirm.

Yep! `@external` as really only allowed on memories (and registers?) anyway.

> The only registering I did for this pass was add it to `default-passes.rs`. Is that all that needs to be done?

It sure seems like it to me!

File: calyx-opt/src/passes/external_to_ref.rs (hunk 1 above)
Comment (nathanielnrn) [reply calyx-opt/src/passes/external_to_ref.rs:32]: cc @rachitnigam When you have a moment, can you confirm if this looks right?

**Comment by ayakayorihiro** [comment]
This is awesome, thank you!! I'll use this pass in my fud2 pipeline :slightly_smiling_face:

**Pull request merged**